freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java

```diff
@@ -0,0 +1,261 @@
+        testFlow(CALLED_ONCE_TILL_REMOVE_MINA, false, false);
+    }
+
+    private void testFlow(int calledOnceTillUpgrade, boolean ccmUpgradeSuccess, boolean userDataUpdateSuccess) throws Exception {
+        FlowIdentifier flowIdentifier = triggerFlow();
+        letItFlow(flowIdentifier);
+
+        flowFinishedSuccessfully(ccmUpgradeSuccess ? 2 : 1);
+        verifyServiceCalls(calledOnceTillUpgrade);
+        verifyFinishingStatCalls(ccmUpgradeSuccess, userDataUpdateSuccess);
+    }
+
+    private void verifyFinishingStatCalls(boolean ccmUpgradeSuccess, boolean userDataUpdateSuccess) throws Exception {
+        verify(upgradeCcmService, times(ccmUpgradeSuccess ? 1 : 0)).finishedState(STACK_ID);
+        verify(resourcesApi, times(userDataUpdateSuccess ? 1 : 0)).updateUserData(any(), any(), any(), eq(USER_DATA));
+
+        verify(upgradeCcmService, times(ccmUpgradeSuccess ? 0 : 1)).failedState(STACK_ID);
+
+        verify(operationService, times(ccmUpgradeSuccess && userDataUpdateSuccess ? 1 : 0)).completeOperation(any(), any(), any(), any());
+        verify(operationService, times(ccmUpgradeSuccess && userDataUpdateSuccess ? 0 : 1)).failOperation(any(), any(), any());
+
+    }
+
+    private void verifyServiceCalls(int tillInt) throws Exception {
+        final int[] expected = new int[ALL_CALLED_ONCE];
+        Arrays.fill(expected, 0, tillInt, 1);
+        int i = 0;
+        verify(upgradeCcmService, times(expected[i++])).checkPrerequisitesState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).checkPrerequsities(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).changeTunnelState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).changeTunnel(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).pushSaltStatesState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).pushSaltStates(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).upgradeState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).upgrade(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).reconfigureState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).reconfigure(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).registerCcmState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).registerCcm(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).healthCheckState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).healthCheck(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).removeMinaState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).removeMina(STACK_ID);
+        verify(userDataService, times(expected[i++])).createUserData(STACK_ID);
+        verify(resourcesApi, times(expected[i++])).updateUserData(any(), any(), any(), eq(USER_DATA));
+
+
+    }
+
+    private void flowFinishedSuccessfully(int numberOfFlows) {
```

Comment:
`numberOfFlows` -> `numberOfSuccessfulFlows`.

Reply:
not necessary successful, rather running. (if first failed, then it has 2 records, but the second flow won't start)

Reply:
Oh, right. Makes sense.